### ISSUE TITLE
No boot screen in cMP 3.1 and RX570 Nitro Sapphire

### ISSUE THREAD

**Issue by jxqqk** [dakanji/RefindPlus#153]
### RefindPlus Version

v0.13.3.AC Release

### Device Type

Apple Mac

### Problem Description

Hi have a cMP 3.1 with a GTX 680 mac edition and a cMP5.1 with a RX 570, both works ok.
I switched the GTX 680 mac edition with a RX 570 (identical to the other one).
Now I have no RefindPlus boot screen, no OC boot screen, but gui works ok.
I tried alternate VBIOS settings, I have RefindPlus boot screen, OC boot screen but black gui.
The VBIOS was modded for mining in both cards in the same way, if I remember well the modded is the alternate version.
I switched cards from cMP 3.1 to cMP 5.1, no changes.
On the cMP 5.1 both VBIOS settings works ok (boot screens and system).
Thank you

### Problem Point

Other

### Affected Items

Other Item

### Debug Log

[cMP51v1.log](https://github.com/dakanji/RefindPlus/files/10772576/cMP51v1.log)
[cMP51v2.log](https://github.com/dakanji/RefindPlus/files/10772577/cMP51v2.log)
[cMP31v1.log](https://github.com/dakanji/RefindPlus/files/10772578/cMP31v1.log)
[cMP31v2.log](https://github.com/dakanji/RefindPlus/files/10772579/cMP31v2.log)

**Comment by startergo** [comment]
> The VBIOS was modded for mining in both cards in the same way,
That is a major no no for the boot screen.

**Comment by jxqqk** [comment]
> > The VBIOS was modded for mining in both cards in the same way,
> 
> That is a major no no for the boot screen. Usually, these changes mess up the CRC check or the signature and make the GOP invalid especially on Nvidia cards.

Only one of the 2 VBIOS available on the cards was modded, the other is original (unfortunately I do not remember witch one).

**Comment by dakanji** [comment]
Please try the MP31 in `TextOnly` mode and share the log.
![Screen Shot 2023-02-18 at 10 17 18](https://user-images.githubusercontent.com/1128876/219847224-ce54a6a1-34f3-4b4b-a3f5-3d00cd2fb2ae.jpg)
You can do this by adding `textonly` to the custom.conf file

It probably will not work in the V1 position (I assume this is the Mining VBios ... please confirm) but it might work in the V2 position.

**Comment by jxqqk** [comment]
> Please try the MP31 in `TextOnly` mode and share the log. ![Screen Shot 2023-02-18 at 10 17 18](https://user-images.githubusercontent.com/1128876/219847224-ce54a6a1-34f3-4b4b-a3f5-3d00cd2fb2ae.jpg) You can do this by adding `textonly` to the custom.conf file
> 
> It probably will not work in the V1 position (This might be the Mining VBios) but it might work in the V2 position.

Same behavior also in textonly:
V1 position: no RF boot screen, no OC boot screen, no apple and progress bar, login screen and gui OK
V2 position: RF boot screen ok, OC boot screen ok, apple and progress bar ok, no login screen and gui (black screen)

**Comment by dakanji** [comment]
Thanks.

Please try the attached build
[X271-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/10773781/X271-BOOTx64.zip)

**Comment by jxqqk** [comment]
V1: no changes
V2: no changes (RF boot screen smaller)
[exp-v1.log](https://github.com/dakanji/RefindPlus/files/10774349/exp-v1.log)
[exp-v2.log](https://github.com/dakanji/RefindPlus/files/10774350/exp-v2.log)

Thank you

**Comment by dakanji** [comment]
Thanks

What is happening is that an RX 5xx will show a screen on the MP31 OOB but needs a certain hack to do the same on MP51. What your unit is showing is that this hack may allow GPUs with "Invalid VBios" to work on MP51. As the hack is not applied on MP31, such GPUs with "Invalid VBios" do not work on MP31.

Trying to find a way to identify this situation and possibly trigger the hack on units, such as MP31, that do not ordinarily need it but only when dealing with such a GPU.

Will need a few rounds to nail it down, if possible, starting with X272 below!
[X272-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/10774598/X272-BOOTx64.zip)

**Comment by jxqqk** [comment]
No changes
[X272v1.log](https://github.com/dakanji/RefindPlus/files/10775304/X272v1.log)

**Comment by dakanji** [comment]
Okay. V1 is a write off and sure that is the case for the MP51 as well. Please focus on V2 instead.

**Comment by dakanji** [comment]
> Please focus on V2 instead.

Actually, still run both for X273
[X273-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/10776046/X273-BOOTx64.zip)

Still need to check on something in V1

Thanks

**Comment by jxqqk** [comment]
No changes
 
[X273v1.log](https://github.com/dakanji/RefindPlus/files/10776324/X273v1.log)
[X273v2.log](https://github.com/dakanji/RefindPlus/files/10776326/X273v2.log)

**Comment by dakanji** [comment]
Thanks. Will soon reach a conclusion.

To keep you up to speed, it turns out the the GPU, assumed to be a "GOP-Only" actually has a UGA instance present and is actually a hybrid GOP/UGA unit. This UGA instance has been confusing things but may be an opportunity although it has a low resolution.

Anyway, please try X274 on both again.
[X274-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/10776384/X274-BOOTx64.zip)

Also, please summarise the result of this test as you did here:  https://github.com/dakanji/RefindPlus/issues/153#issuecomment-1435527671

Thanks

**Comment by jxqqk** [comment]
Ok, X274 run:
V1 position: no RF boot screen, no OC boot screen, no apple and progress bar, login screen and gui OK
V2 position: RF boot screen ok, OC boot screen ok, apple and progress bar ok, no login screen and gui (black screen)
[X274v1.log](https://github.com/dakanji/RefindPlus/files/10777169/X274v1.log)
[X274v2.log](https://github.com/dakanji/RefindPlus/files/10777173/X274v2.log)

**Comment by dakanji** [comment]
Thanks.

Can you test on the MP51 and similarly report?

**Comment by jxqqk** [comment]
X274 run on MP51:
V1 position: RF boot screen OK, OC boot screen OK, OS boot OK (apple, progress bar and gui)
V2 position: RF boot screen OK, OC boot screen OK, OS boot OK (apple, progress bar and gui)
[51-X274v1.log](https://github.com/dakanji/RefindPlus/files/10777357/51-X274v1.log)
[51-X274v2.log](https://github.com/dakanji/RefindPlus/files/10777358/51-X274v2.log)

**Comment by dakanji** [comment]
I know you have previously said that the two GPUs were treated the same way but can you actually physically swap them?

There is no reason I can think of why they should be acting differently apart from that they might be in fact different in some way.

They are acting differently right off the bat, before any hack for the MP51 takes place and this suggests that they might be different.

**Comment by dakanji** [comment]
Please try X275 on the MP31.
[X275-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/10777400/X275-BOOTx64.zip)

Thanks

**Comment by jxqqk** [comment]
No changes:
[X275v1.log](https://github.com/dakanji/RefindPlus/files/10777559/X275v1.log)
[X275v2.log](https://github.com/dakanji/RefindPlus/files/10777560/X275v2.log)

I modded the cards some years ago, and I think I modded v2 but I can not be sure about anything.
Both are same model, but I can not remember if they have the same memory chips, if this is the case they can have different timings.
When I tried the MP51 card on the MP31, I tried only v1 and I had exactly the same behaviour, no boot screens but gui ok.

**Comment by dakanji** [comment]
I don't think it is a timing issue. 

In summary, RefindPlus, and OpenCore it seems, is unable to detect GOP on that GPU when it is running on MP31 in the V1 position at all. The Mac OS kernel is however apparently able to both detect and activate this GOP. As of X274, RefindPlus is able to "properly" handle that in the V2 position but Mac OS kernel then apparently has an issue. On the MP51 on the other hand, everything apparently works as normal. 

I can't wrap my head around it.

To get a baseline for moving forward, use OpenCore directly by cutting RefindPlus out of the boot chain and confirm whether the behaviour is as above. That is, eliminate or confirm RefindPlus as a factor on both units with the GPU VBios in both settings. You might want to swap between both machines as well. That is, the MP51 GPU in the MP31 and vice versa. Again in both positions.

Actually, you might want to do the full swap test with RefindPlus first. You can ditch X275 and go back to X274.

Thanks

**Comment by jxqqk** [comment]
Thanks.
The MP51 is still a production machine, I need some day to try the swap.
The trial seems to say the original VBIOS is v2, in the meantime I try to focus on that position with RF and OC boot working and find the macos kernel issue strangely not present in the v1, I can try to tweak the OC instance.

**Comment by dakanji** [comment]
Suggestion was:
WIth RefindPlus X274 (Actually X275 is fine)
- Shut both units down (I assume MP51 does not have to be on 24/7)
- Put the MP31 GPU into the MP51
  - Start the MP51 in V1 and note what happens
  - Start the MP51 in V2 and note what happens
  - Shut MP51 down
- Put the MP51 GPU into the MP31
  - Start the MP31 in V1 and note what happens
  - Start the MP31 in V2 and note what happens
  - Shut MP31 down
- Restore the GPUs back as they were

Later on, disable RefindPlus so that OpenCore is loaded directly and repeat to see if the results are the same.

**Comment by dakanji** [comment]
Right, I have spent some time going over things in quite a bit of detail and reached the conclusion that whatever is behind the issues is not RefindPlus related or at least, not materially so. While you have not been able to run the exclusion tests, you will need to look into the GPU or OpenCore config or of course the MP31 capabilities. If I had to take a punt, I would say it is down to the OpenCore config.

Anyway, I will have to tag this as "Invalid" but it has been a very useful exercise nonetheless and I have been able to improve the overall graphics handling in RefindPlus based on the multiple tests you helped with as a result.

Thanks!
